---
title: 'ATAC-seq: differential accessability analysis'
author: "Sam Buckberry"
output: html_document
---
---
title: "Untitled"
output: html_document
---

```{r}
# BioC packages
library(GenomicRanges)
library(GenomicAlignments)
library(edgeR)
library(ggbio)

# CRAN packages
library(stringr)
library(magrittr)
library(ggplot2)
library(ggrepel)
library(pheatmap)
library(tidyr)
```

https://asia.ensembl.org/info/genome/funcgen/regulatory_build.html

http://ftp.ensembl.org/pub/current_regulation/homo_sapiens/homo_sapiens.GRCh38.Regulatory_Build.regulatory_features.20210107.gff.gz

Read the peak counts
```{r}
peak_counts <- read.table("processed_data/atac_peak_counts.tsv")
```

Set the treatment and control groups
```{r}
groups <- ifelse(test = grepl(pattern = "_T.U", x = colnames(peak_counts)),
                 yes = "Unstimulated", no = "Stimulated")
samples <- str_sub(string = colnames(peak_counts), start = 2, end = 5)
samples <- str_c(samples, groups, sep = "_")

```
Load counts into a DGEList
```{r}
y <- DGEList(counts = peak_counts, group = groups, samples = samples)
```

Design matrix for differential testing
```{r}
design <- model.matrix(~ 0 + groups)
design

# Clean up the column names (will make things easier later)
colnames(design) <- c("Stimulated", "Unstimulated")
design
```


Filter to remove peaks with low counts that would not be useful in statistical testing
```{r}
keep <- filterByExpr(y)
table(keep)

y <- y[keep, , keep.lib.sizes=FALSE]
```

Calculate normalisation factors
```{r}
y <- calcNormFactors(y)
y$samples
```


Differential testing using Limma-Voom
```{r}
v <- voom(counts = y, design = design, plot = TRUE)
```

What is voom doing?

Counts are transformed to log2 counts per million reads (CPM), where “per million reads” is defined based on the normalization factors we calculated earlier.
A linear model is fitted to the log2 CPM for each peak, and the residuals are calculated.
A smoothed curve is fitted to the sqrt(residual standard deviation) by average peak counts (see red line in plot above)
The smoothed curve is used to obtain weights for each peak and sample that are passed into limma along with the log2 CPMs.
More details for the application of this method to RNA-seq data can be found here.  https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29


```{r, fig.height=5}
boxplot(v$E, names=v$targets$samples)
```

```{r}
plotMDS(v$E)
```



If the boxplots look markedly different between samples (the above look ok), a quantile normalisation might be beneficial.
```{r}
vq <- voom(counts = y, design = design, plot = TRUE,
           normalize.method = "quantile")
```

```{r, fig.height=5}
boxplot(vq$E, names=v$targets$samples)
```

```{r}
plotMDS(vq$E)
```

### Differential expression testing
Fitting linear models in Limma
```{r}
fit <- lmFit(object = v, design = design)
```

Make the contrasts matrix
```{r}
contr <- makeContrasts(Stimulated - Unstimulated, levels = colnames(coef(fit)))
contr
```

Estimate contrasts for each gene
```{r}
fit2 <- contrasts.fit(fit, contr)
```

Empirical Bayes smoothing of standard errors (shrinks standard errors that are much larger or smaller than those from other peaks towards the average standard error) (see https://www.degruyter.com/doi/10.2202/1544-6115.1027)
```{r}
eb <- eBayes(fit2)
```

Check out the top differential peaks
```{r}
topTable(eb)
```

Put all DE testing results into an object for plotting and saving
```{r}
tt2 <- topTable(eb, number = nrow(v))
tt2$Significant <- ((abs(tt2$logFC) > 1) & (tt2$adj.P.Val < 0.05))
```

How many DE peaks are there?
```{r}
table(tt2$Significant)
```

### Differential testing summary plots

Volcano plot
```{r}
gg_volcano <- ggplot(tt2, aes(x = logFC, y = -log10(P.Value),
                             fill=Significant, colour=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(-1, 1), linetype="dashed")
gg_volcano
```

MA plot
```{r}
gg_ma <- ggplot(tt2, aes(x = AveExpr, y = logFC, colour=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(1), linetype="dashed") +
    geom_hline(yintercept = c(-1, 1), linetype="dashed")
gg_ma
```

### Significant peak plots

DE gene heatmap of DE genes
```{r}
top_de_cpm <- v$E[rownames(v$E) %in% rownames(tt2)[tt2$Significant == TRUE], ]
colnames(top_de_cpm) <- v$targets$samples

pheatmap(top_de_cpm, scale = "row", show_rownames = FALSE)
```

Plot normalised peak counts for top peaks
```{r}
plot_de <- function(ids){
    
    # Get the normalised expression data for the selected genes
    dat <- v$E[rownames(v$E) %in% ids, ] %>% 
        data.frame()
    colnames(dat) <- v$targets$samples
    
    # Re-shape the data into the ggplot preferred long format
    dat <- tibble::rownames_to_column(dat, var = "id") %>%
        tidyr::gather(sample, value, -id)
    
    # Add the group data
    dat$group <- v$targets$group[match(dat$sample, v$targets$samples)]
    
    # plot the gene expression vaues
    ggplot(data = dat, aes(x = group, y = value)) + 
        geom_point(size=3, alpha=0.5) +
        facet_wrap(~id, scales = "free_y") +
        ylab("Normalised counts") +
        theme_bw()
        
}

my_top_peaks <- rownames(tt2)[tt2$Significant == TRUE][1:6]

plot_de(ids = my_top_peaks)
```

### Annotate differential peaks
Get the significant peaks up- and down-regulated into GRanges objects
```{r}
sig_peaks_up <- rownames(tt2)[tt2$Significant == TRUE & tt2$logFC > 0] %>%
    GRanges()

sig_peaks_down <- rownames(tt2)[tt2$Significant == TRUE & tt2$logFC < 0] %>%
    GRanges()
```

Import ENCODE regulatory elements data
```{r}
reg_gr <- rtracklayer::import("reference/homo_sapiens.GRCh38.Regulatory_Build.regulatory_features.20210107.gff.gz", format = "gff")
```

Get the overlapping regulatory elements for the peaks
```{r}

annotate_peaks <- function(query_gr, subject_gr, subject_annotate_col="feature_type"){
    
    # Get overlapping ranges
    olaps <- findOverlaps(query = query_gr, subject = subject_gr)

    # Get query ranges with no hits
    gr_no_hit <- query_gr[!overlapsAny(query_gr, subject_gr)]    
    gr_no_hit$feature <- "No feature"
    
    # Get query ranges with hits.
    # Note that this permits query ranges that overlap more than one subject range
    
    gr_hit <- query_gr[olaps@from]
    gr_hit$feature <- mcols(subject_gr)[olaps@to, subject_annotate_col]
    
    gr <- c(gr_no_hit, gr_hit)
    
    return(gr)
}

peaks_up_annot <- annotate_peaks(query_gr = sig_peaks_up, subject_gr = reg_gr)
peaks_up_annot$direction <- "Up"

peaks_down_annot <- annotate_peaks(query_gr = sig_peaks_down, subject_gr = reg_gr)
peaks_down_annot$direction <- "Down"

all_peaks_annot <- c(peaks_up_annot, peaks_down_annot) %>% as.data.frame()
head(all_peaks_annot)
```

Plot the peak annotations
```{r}
gg_annot <- ggplot(all_peaks_annot, aes(x = direction, fill=feature)) +
    geom_bar() + theme_bw()

gg_annot
```

...or plot proportionally
```{r}
gg_annot_prop <- ggplot(all_peaks_annot, aes(x = direction, fill=feature)) +
    geom_bar(position = "fill") + 
    geom_text(
    aes(label=signif(..count.. / tapply(..count.., ..x.., sum)[as.character(..x..)], digits=2)),
    stat="count",
    position=position_fill(vjust=0.5)) +
    labs(y="Proportion") +
    theme_bw()

gg_annot_prop

```

