---
title: "Untitled"
author: "Sam Buckberry"
date: "20/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(GenomicRanges)
library(GenomicAlignments)
```


### Post-alignment processing
Much of the post-alignment processing is done with `samtools` outside of R. While some of this is possible with the `Rsamtools` package, it is much more efficient to write a short shell script with the following steps. 

- Filter reads of mitochondrial data
- Filter for 'proper' read pairs
- Sort the alignments
- Index the alignments
- Remove PCR duplicates

Here are the lines of code for each step:

Remove mitochondrial mapped reads
```{bash, eval=FALSE}
samtools idxstats my_atac.bam | cut -f 1 | grep -v chrM | xargs samtools view -b my_atac.bam > my_atac_filtered.bam

```

Remove reads mapped to blacklisted regions
http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/

Sort the alignments
```{bash, eval=FALSE}
samtools sort my_atac_filtered.bam > my_atac_filtered_sorted.bam
```

Index the sorted alignments
```{bash, eval=FALSE}
samtools index my_atac_filtered_sorted.bam
```

Remove PCR duplicates
```{bash, eval=FALSE}
samtools dedup my_atac_filtered_sorted.bam > my_atac_dedup.bam
```

And we can put that all together in a shell script like this
```{bash, eval=FALSE}
#!/bin/bash

# File inputs
bam=$1
blacklist=$2

prefix=$(basename "$bam" .bam)

# Filter for nuclear genome
samtools idxstats "$bam" | cut -f 1 | grep -v \* | grep -v MT | grep -v chrM | xargs samtools view -b "$bam" > "$prefix"_nuclear.bam

# Remove blacklisted mapped reads
samtools view -L "$blacklist" -U "$prefix"_filtered.bam -b "$prefix"_nuclear.bam > "$prefix".blacklist.bam

# Sort the BAM file
samtools sort "$prefix"_filtered.bam > "$prefix"_sorted.bam

# Index the sorted alignments
samtools index "$prefix"_sorted.bam

# Remove the PCR duplicates
samtools rmdup "$prefix"_sorted.bam "$prefix"_dedup.bam

# Index the dedup BAM file
samtools index "$prefix"_dedup.bam

# Clean up the intermediate files
rm "$prefix"_filtered.bam "$prefix"_sorted.bam "$prefix"_sorted.bam.bai "$prefix".blacklist.bam 

```

Process all alignments with the bash script
```{bash}


```


Make Tn5 insertion-centered bed files for peak calling, and bigwig files for browser and plotting. 
```{r}

process_atac_bam <- function(bam){
    
    # Get single reads loaded for offsetting to get tn5 insertion site
    message(Sys.time())
    
    # Read the BAM file in 10,000 read chunks to preserve memory
    message(str_c("Reading ", bam))
    bam_file <- BamFile(bam, yieldSize = 10000)
    open(bam_file)
    bam_gr <- readGAlignments(file = bam) %>% GRanges()
    close(bam_file)

    message("Offset alignments to Tn5 insertion sites")
    pos <- bam_gr[strand(bam_gr) == "+"] %>% 
        GenomicRanges::shift(shift=4) %>%
        resize(width = 50, fix = "start")
    
    neg <- bam_gr[strand(bam_gr) == "-"] %>%
        GenomicRanges::shift(shift = -5) %>%
        resize(width = 50, fix = "start")
    
    shift_gr <- c(pos, neg)
    rm(pos, neg, bam_gr); gc()
    strand(shift_gr) <- "+"
    
    # Calculate Tn insertion coverage
    message("Calculating coverage")
    cov <- coverage(shift_gr)
    
    # Normalise by sequence depth (CPM)
    message("Normalising by depth")
    lib_size <- length(shift_gr)/1e6
    cov_cpm <- cov / lib_size
    
    # write tn5 centered atac bigwig
    out_bw <- str_replace(string = bam, pattern = ".bam",
                          replacement = "_atac.bw")
    
    message(str_c("Writing ", out_bw))
    
    export.bw(object = cov_cpm, con = out_bw)
    
    # write the tn5 centered bed file for peak calling
    dat <- as.data.frame(shift_gr)[ ,1:3]
    dat$name <- "."
    dat$score <- "."
    dat$strand <- "+"
    
    out_bed <- str_replace(string = bam, pattern = ".bam",
                           replacement = "_atac.bed")
    
    message(str_c("Writing ", out_bed))
    write.table(x = dat, file = out_bed, quote = FALSE, sep = "\t",
                        row.names = FALSE, col.names = FALSE)

    message(str_c("Completed ", Sys.time()))
        
}

fls <- list.files("aligned_data/", pattern = "chr22.bam$",
                       full.names = TRUE)

bam_list <- BamFileList(fls, yieldSize = 10000, asMates = FALSE)

process_atac_bam(bam_list[[3]])

lapply(bam_list, process_atac_bam)
```

Call peaks for each sample using macs2
```{r}

atac_bed <- "aligned_data/test.bed"

call_macs_peaks <- function(atac_bed){
    
    macs_name <- basename(atac_bed) %>% 
        str_remove(pattern = ".bed") %>%
        str_c("peaks", .)
    
    cmd <- str_c("macs2 callpeak --gsize hs --nomodel --keep-dup all --name ",
                 macs_name, " -t ", atac_bed)
    
    system(cmd)
    
}

# List the atac bed files
atac_bed_list <- list.files("aligned_data/",
                            pattern = ".bed$", full.names = TRUE)

# Run the peak calling function for all of the bed files



```








